File: roverProblem.py
#string input

#---- Testing Functions -----
def areNumbers(testList):
    if len(testList) == 0: return False

    for element in testList:
        if(not element.isdigit()): return False

    return True


def isValidCoordinate(testDirection):
    if testDirection == "N" or testDirection == "S" or testDirection == "W" or testDirection == "E":
        return True
    return False


def isValidRoverPosition(x, y, direction):
    if(areNumbers((x,y)) and isValidCoordinate(direction)): 
        return True
    return False

def isValidRoverInitialPosition(x, y, direction, xFinal, yFinal):
    if isValidRoverPosition(x, y, direction):
        if 0 <= int(x) <= xFinal and 0 <= int(y) <= yFinal:
            return True
    return False

#------ Error handlers -----
def errorPlateuSize(plateauSize):
    if len(plateauSize) < 2 or not areNumbers(plateauSize):
        print("Error: Invalid coordinates for Plateu, try again")
        return True
    return False

def errorListIndex(index, testingList):
    if index+1 > len(testingList):
        print("Error: No instructions founded for rover in line: " + index)
        return True
    return False

def errorRoverCoordinate(testingList):
    if len(testingList) < 3:
        print("Error: Insuficient values for roger, needs (x, y, direction) got: " + " ".join([str(x) for x in testingList]))
        return True
    elif not isValidRoverPosition(testingList[0], testingList[1], testingList[2]):
        print("Error: Invalid values for roger, needs [x(number), y(number), direction(N/S/E/W)]. got: " + " ".join([str(x) for x in testingList]))
        return True
    return False



#----- Main Functions ----


def moveRover(x, y, orientation, limitX, limitY, instructions):
    currentX = x
    currentY = y
    currentOrientation = orientation
    for instruction in instructions.upper():
        if(instruction == "M"):
            if(currentOrientation == "N" and currentY < limitY):
                currentY+=1
            elif(currentOrientation == "S" and currentY > 0):
                currentY-=1
            elif(currentOrientation == "E" and currentX < limitX):
                currentX+=1
            elif(currentOrientation == "W" and currentX > 0):
                currentX-=1
        elif(instruction == "L"):
            match currentOrientation:
                case "N":
                    currentOrientation = "W"
                case "W":
                    currentOrientation = "S"
                case "S":
                    currentOrientation = "E"
                case "E":
                    currentOrientation = "N"
        elif(instruction == "R"):
            match currentOrientation:
                case "N":
                    currentOrientation = "E"
                case "W":
                    currentOrientation = "N"
                case "S":
                    currentOrientation = "W"
                case "E":
                    currentOrientation = "S"
        else:
            print("Error: Invalid instruction in string founded: " + instruction+ ", canceling movement")
            return [x, y, orientation]
    return [currentX, currentY, currentOrientation]



def roversPosition(instructions):

    results = []
    filteredInputList = list(filter(lambda x: len(x) > 0, instructions.split("\n")))
    plateauSize = filteredInputList[0].strip().split(" ")[0:2]

    if errorPlateuSize(plateauSize): return results
    [xLimit, yLimit] = map(lambda coordinate: int(coordinate), plateauSize)
    
    for index in range(1, len(filteredInputList), 2):

        initialPosition = filteredInputList[index].strip().upper().split()

        if errorRoverCoordinate(initialPosition): continue
        [xPosition, yPosition, direction] = initialPosition[0:3]

        if(isValidRoverInitialPosition(xPosition, yPosition, direction, xLimit, yLimit)):
            if errorListIndex(index, filteredInputList): return results
            instructionList = filteredInputList[index+1]
            
            roverFinalPosition = moveRover(int(xPosition), int(yPosition), direction, xLimit, yLimit, instructionList.strip())
            stringFinalPosition = " ".join([str(x) for x in roverFinalPosition])
            results.append(stringFinalPosition)

        else:
            print("Invalid rover position, continuing with next rover")
    return results
        
        
stringInput = """
5 5
1 2 N 
LMLMLMLMM 
5 3 E
MMRMMRMRRM
"""

roversEndingPositions = roversPosition(stringInput)

for roverPosition in roversEndingPositions:
    print(roverPosition)
print("==========")

File: roverProblem.php
<?php

// Testing Functions
function areNumbers($testList) {
    
    if (count($testList) == 0) return false;

    foreach ($testList as $element) {
        if (!ctype_digit($element)) return false;
    }

    return true;
}

function isValidCoordinate($testDirection) {
    return $testDirection == "N" || $testDirection == "S" || $testDirection == "W" || $testDirection == "E";
}

function isValidRoverPosition($x, $y, $direction) {
    return areNumbers([$x, $y]) && isValidCoordinate($direction);
}

function isValidRoverInitialPosition($x, $y, $direction, $xFinal, $yFinal) {
    if (isValidRoverPosition($x, $y, $direction)) {
        if (0 <= (int)$x && (int)$x <= $xFinal && 0 <= (int)$y && (int)$y <= $yFinal) {
            return true;
        }
    }
    return false;
}

// Error handlers
function errorPlateuSize($plateauSize) {
    if (count($plateauSize) < 2 || !areNumbers($plateauSize)) {
        echo "Error: Invalid coordinates for Plateu, try again\n";
        return true;
    }
    return false;
}

function errorListIndex($index, $testingList) {
    if ($index + 1 > count($testingList)) {
        echo "Error: No instructions founded for rover in line: " . $index . "\n";
        return true;
    }
    return false;
}

function errorRoverCoordinate($testingList) {
    if (count($testingList) < 3) {
        echo "Error: Insuficient values for roger, needs (x, y, direction) got: " . implode(" ", $testingList) . "\n";
        return true;
    } elseif (!isValidRoverPosition($testingList[0], $testingList[1], $testingList[2])) {
        echo "Error: Invalid values for roger, needs [x(number), y(number), direction(N/S/E/W)]. got: " . implode(" ", $testingList) . "\n";
        return true;
    }
    return false;
}

// Main Functions
function moveRover($x, $y, $orientation, $limitX, $limitY, $instructions) {
    $currentX = $x;
    $currentY = $y;
    $currentOrientation = $orientation;
    $instructions = strtoupper($instructions);

    for ($i = 0; $i < strlen($instructions); $i++) {
        $instruction = $instructions[$i];
        if ($instruction == "M") {
            if ($currentOrientation == "N" && $currentY < $limitY) {
                $currentY++;
            } elseif ($currentOrientation == "S" && $currentY > 0) {
                $currentY--;
            } elseif ($currentOrientation == "E" && $currentX < $limitX) {
                $currentX++;
            } elseif ($currentOrientation == "W" && $currentX > 0) {
                $currentX--;
            }
        } elseif ($instruction == "L") {
            switch ($currentOrientation) {
                case "N":
                    $currentOrientation = "W";
                    break;
                case "W":
                    $currentOrientation = "S";
                    break;
                case "S":
                    $currentOrientation = "E";
                    break;
                case "E":
                    $currentOrientation = "N";
                    break;
            }
        } elseif ($instruction == "R") {
            switch ($currentOrientation) {
                case "N":
                    $currentOrientation = "E";
                    break;
                case "W":
                    $currentOrientation = "N";
                    break;
                case "S":
                    $currentOrientation = "W";
                    break;
                case "E":
                    $currentOrientation = "S";
                    break;
            }
        } else {
            echo "Error: Invalid instruction in string founded: " . $instruction . ", canceling movement\n";
            return [$x, $y, $orientation];
        }
    }
    return [$currentX, $currentY, $currentOrientation];
}

function roversPosition($instructions) {
    $results = [];
    $filteredInputList = array_values(array_filter(explode("\n", $instructions), function($el) {
        return strlen(trim($el)) > 0;
    }));
    
    $plateauSize = array_slice(explode(" ", trim($filteredInputList[0])), 0, 2);

    if (errorPlateuSize($plateauSize)) return $results;

    [$xLimit, $yLimit] = array_map('intval', $plateauSize);

    for ($index = 1; $index < count($filteredInputList); $index += 2) {
        $initialPosition = explode(" ", strtoupper(trim($filteredInputList[$index])));
        if (errorRoverCoordinate($initialPosition)) continue;

        [$xPosition, $yPosition, $direction] = array_slice($initialPosition, 0, 3);

        if (isValidRoverInitialPosition($xPosition, $yPosition, $direction, $xLimit, $yLimit)) {
            if (errorListIndex($index, $filteredInputList)) return $results;

            $instructionList = trim($filteredInputList[$index + 1]);
            $roverFinalPosition = moveRover((int)$xPosition, (int)$yPosition, $direction, $xLimit, $yLimit, $instructionList);
            $stringFinalPosition = implode(" ", $roverFinalPosition);
            $results[] = $stringFinalPosition;
        } else {
            echo "Invalid rover position, continuing with next rover\n";
        }
    }
    return $results;
}

$stringInput = "
5 5
1 2 N 
LMLMLMLMM 
5 3 E
MMRMMRMRRM
";

$roversEndingPositions = roversPosition($stringInput);

foreach ($roversEndingPositions as $roverPosition) {
    echo $roverPosition . "\n";
}
echo "==========\n";
?>